hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/savepoint/SavepointActionExecutor.java

```diff
@@ -145,10 +132,41 @@ public HoodieSavepointMetadata execute() {
           .saveAsComplete(
               true, instantGenerator.createNewInstant(HoodieInstant.State.INFLIGHT, HoodieTimeline.SAVEPOINT_ACTION, instantTime), Option.of(metadata),
               savepointCompletedInstant -> table.getMetaClient().getTableFormat().savepoint(savepointCompletedInstant, table.getContext(), table.getMetaClient(), table.getViewManager()));
-      LOG.info("Savepoint " + instantTime + " created");
+      LOG.info("Savepoint {} created", instantTime);
       return metadata;
     } catch (HoodieIOException e) {
       throw new HoodieSavepointException("Failed to savepoint " + instantTime, e);
     }
   }
+
+  @VisibleForTesting
+  String getLastCommitRetained() {
+    Option<HoodieInstant> cleanInstant = table.getCleanTimeline().lastInstant();
+    // if there are no clean instants, we can use the first completed commit
+    if (cleanInstant.isEmpty()) {
+      return table.getCompletedCommitsTimeline().firstInstant().get().requestedTime();
+    }
+    return cleanInstant.map(instant -> {
+      try {
+        if (instant.isCompleted()) {
+          return table.getActiveTimeline().readCleanMetadata(instant)
+              .getEarliestCommitToRetain();
+        } else {
+          // clean is pending or inflight
+          return table.getActiveTimeline().readCleanerPlan(
+                  instantGenerator.createNewInstant(REQUESTED, instant.getAction(), instant.requestedTime()))
+              .getEarliestInstantToRetain().getTimestamp();
+        }
+      } catch (IOException e) {
+        throw new HoodieSavepointException("Failed to savepoint " + instantTime, e);
+      }
+    }).orElseGet(() ->
+      // If there is no earliest commit to retain, but there are clean instants on the timeline, we must assume KEEP_LATEST_FILE_VERSIONS policy is used.
+      // In that case, the last commit before the clean instant is the last commit guaranteed to be retained.
+      table.getActiveTimeline().getWriteTimeline().filterCompletedInstants()
```

Comment:
if earliest commit to retain is not found, can we add a warn log to recommend users to switch to KEEP_LATEST_COMMITS or KEEP_LATEST_BY_HOURS if they are looking to savepoint earlier commits. 
we can still proceed w/ whats best we could do for current savepoint trigger. 
but user can switch the clean policy from next time.

Reply:
We already have validation check in L82.